tools/shared.py

```diff
@@ -1743,8 +1743,15 @@ def link(files, target, force_archive_contents=False, temp_files=None, just_calc
     # returns True.
     def consider_object(f, force_add=False):
       new_symbols = Building.llvm_nm(f)
-      if not new_symbols.is_valid():
-        logging.warning('object %s is not valid, cannot link' % (f))
+      # Check if the object was valid according to llvm-nm. It also accepts
+      # native object files.
+      if not new_symbols.is_valid_for_nm():
+        logging.warning('object %s is not valid according to llvm-nm, cannot link' % (f))
```

Comment:
Why aren't these just errors?  In fact, why doesn't Building.llvm_nm() just raise an exception in this case rather than returning an empty ObjectInfo?

Reply:
Sometimes .a files contain extra things, like Rust places metadata files in them. We need to just skip over those, basically, so we can link in the relevant code.

Another option here might be to only warn if all the contents of a .a are invalid. That would point out .a files that contain native object files instead of LLVM/wasm ones. I'm not sure that's enough, though, my guess is the extra warnings here (on each invalid file) are worth it.

Reply:
ok.. that makes sense